Adds AWS EKS terraform config

jasonhawkharris opened the pull request:
- adds terraform config
- adds/edits all readme's for clarity

jasonhawkharris committed:
initial aws terraform config

jasonhawkharris committed:
AWS final changes and README edits
- Uncomment desired_state variables in eks module
- Reset tfvars to default and comment out
- Add README to AWS folder
- Re route `kubectl` info to root level README
- Adds dev scripts to .gitignore

jasonhawkharris committed:
clarify steps in README

abeatrix commented on aws/variables.tf:
```diff
@@ -0,0 +1,55 @@
+variable "node_group_two_instance_types" {
+  description = "Instance types to use in node group two"
+  type        = list
+  default     = ["m5.4xlarge"]
```
```suggestion
  default     = ["m6a.4xlarge"]
```

abeatrix commented on aws/variables.tf:
```diff
@@ -0,0 +1,55 @@
+variable "node_group_one_instance_types" {
+  description = "Instance types to use in node group two"
+  type        = list
+  default     = ["m5.4xlarge"]
```
```suggestion
  default     = ["m6a.4xlarge"]
```

I think [m6a](https://aws.amazon.com/ec2/instance-types/m6a/) is cheaper with better performance

![image](https://user-images.githubusercontent.com/68532117/213617171-cf21dc2c-19a1-4a04-933d-d6a3c3ff2812.png)

![image](https://user-images.githubusercontent.com/68532117/213617102-5d7f82bf-8d70-4617-8c5b-644854f90d5f.png)

abeatrix commented on aws/eks.tf:
```diff
@@ -0,0 +1,71 @@
+    create_security_group = false
+  }
+
+  eks_managed_node_groups = {
+    one = {
```
What is the second node group for? (I'm new to AWS and terraform so I might have missed something)

daxmc99 replied:
Similar question here, a HA-cluster configuration might want two node groups and we should have a HA cluster setup, but the default can be in a single AWS availability zone.

daxmc99 replied:
This seems to be unresolved

jasonhawkharris replied:
I've removed the second node group in a recent commit, but I left the `one = {}` syntax to make it easier to add more node_groups if the user wants to make the cluster HA.

abeatrix replied:
What would be the terraform setting to enable HA (assuming this is what you want to support based on recent comment) in this case when leaving `one`?  

It might be worth looking into updating this into[ a dynamic block ](https://developer.hashicorp.com/terraform/language/expressions/dynamic-blocks) and creating node groups based on user input instead of hard-coding just one/two if I'm understanding the purpose correctly?

abeatrix commented on aws/README.md:
```diff
@@ -0,0 +1,24 @@
+### Destroy the Cluster
+1. Run `terraform destroy`. Answer yes at the prompt.
+
+Congrats! you have a Kubernetes cluster! Next, [it's time to deploy Sourcegraph](https://docs.sourcegrapn.com/admin/deploy/kubernetes/configure)
```
Maybe move this before the `Destory the Cluster section` 😆 ?

abeatrix reviewed (commented):
Is this supposed to set up a cluster only, without Sourcegraph being deployed?

daxmc99 commented on aws/eks.tf:
```diff
@@ -0,0 +1,71 @@
+      desired_size = var.node_group_one_desired_size
+
+      pre_bootstrap_user_data = <<-EOT
+      echo 'foo bar'
```
Consider removing?

jasonhawkharris replied:
When I remove the `desired_size` variable, I get a failed deployment. When I keep it in there, It works swimmingly and allows me to deploy a Sourcegraph instance with no problem.

What would be the reason for removing `pre_bootstrap_user_data`?

daxmc99 replied:
This explicitly commenting on line 35 `echo 'foo bar'` 
It doesn't seem like we need this
`echo` is a Unix command that prints the string to the terminal. `pre_boostrap_user_data` is usually where you put commands you need to run on the instance being created. EKS nodes usually don't require this

daxmc99 commented on aws/eks.tf:
```diff
@@ -0,0 +1,71 @@
+      desired_size = var.node_group_two_desired_size
+
+      pre_bootstrap_user_data = <<-EOT
+      echo 'foo bar'
```
Same as above, consider removing

daxmc99 commented on aws/terraform.tfvars:
```diff
@@ -0,0 +1,17 @@
+### AWS Region ###
+# region = "us-west-1" 
```
Why is this file commented out?

daxmc99 commented on aws/versions.tf:
```diff
@@ -0,0 +1,15 @@
+# terraform {
+#   required_providers {
```
This file is usually called a `providers.tf` https://docs.sourcegraph.com/dev/background-information/languages/terraform#providers-tf

jasonhawkharris replied:
Good callout. I actually commented this out when I was getting some errors when running. Meant to come back later and update the versions so that it worked, but found that it worked perfectly without this file at all. Should I add it back?

daxmc99 reviewed (changes requested):
Looking good! Nice progress!

jasonhawkharris commented:
@abeatrix Yes. This is for deploying a cluster only. But it deploys a cluster with ideal specs for provisioning a Sourcegraph instance.

jasonhawkharris committed:
change node instance type

jasonhawkharris committed:
remove second node group; readme changesOC

jasonhawkharris committed:
Uncomment and rename versions.tf --> providers.tf

jasonhawkharris commented:
Tested and ready to go with requested changes. Requesting re-review.

jasonhawkharris committed:
formatting

daxmc99 reviewed (changes requested):
Still need resolution on some of the comments

jasonhawkharris committed:
remove pre_bootstrap_user_data

daxmc99 commented on aws/eks.tf:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  eks_managed_node_group_defaults = {
+    ami_type = "AL2_x86_64"
```
Maybe a link to docs in a comment about what AMI this is?

jdpleiness commented on aws/iam.tf:
```diff
@@ -0,0 +1,170 @@
+
+  # Terraform's "jsonencode" function converts a
+  # Terraform expression result to valid JSON syntax.
+  policy = file("./resources/iam-policy.json")
```
Consider replacing reading a json file with a[ data source policy document ](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/iam_policy_document)

jasonhawkharris committed:
add to .gitignore

jasonhawkharris committed:
Change default node pool name; add tf iam iam-policy
- Change name of default node pool to default instead of one.
- Removes resources folder with hardcoded json iam policy and replace
  with iampolicy.tf

daxmc99 commented on README.md:
```diff
@@ -1,3 +1,24 @@
+
+`aws eks update-kubeconfig --region $REGION --name $CLUSTER_NAME`
+
+4. Run `kubectl cluster-info` to test if it worked.
```
Consider replacing with context
```suggestion
4. Run `kubectl config current-context` to see the current context
5. (Optionally) Run `kubectl cluster-info` to see information about the current cluster
```

daxmc99 commented on README.md:
````diff
@@ -1,3 +1,24 @@
+To manage your cluster, you can use the `kubectl` [cli tool](https://kubernetes.io/docs/tasks/tools/), which you'll need to configure cluster access for.
+
+1. For both commands you'll need to set the env variables first:
+    ```
````
```suggestion
    ```bash
```

daxmc99 commented on aws/eks.tf:
```diff
@@ -0,0 +1,44 @@
+
+  eks_managed_node_groups = {
+    default = {
+      name = "node-group-1"
```
```suggestion
      name = "sg-node-group-1"
```

daxmc99 reviewed (approved):
LGTM
I haven't run this myself by high-level the approach looks good

jasonhawkharris committed:
use current-context command instead of cluster-info

Co-authored-by: Dax McDonald <dax@sourcegraph.com>

jasonhawkharris committed:
Update README.md

Co-authored-by: Dax McDonald <dax@sourcegraph.com>

jasonhawkharris committed:
change node group name

Co-authored-by: Dax McDonald <dax@sourcegraph.com>

jasonhawkharris committed:
Update eks.tf
